### DISCUSSION TITLE
Asking for Pro Tips on simple local https deployment - classrooms, small office/home office settings

### DISCUSSION THREAD

**Discussion by tbendien** [open-webui/open-webui#1036] [Ideas]
@justinh-rahb  @tjbck us I am looking to provide small GPU OWUI computers to school classrooms and home offices for use on local wifi networks. Any Pro Tips or suggestions on how to implement basic https in that type of local deployment? A self signed cert should be fine for now, I just want to have the http traffic encrypted while transiting the WiFi or home network.

![IMG_6997](https://github.com/open-webui/open-webui/assets/47579015/f5cacd87-b2bd-4d52-8842-69ae4c1fdf83)
![IMG_6872](https://github.com/open-webui/open-webui/assets/47579015/9472f582-ae6a-4e4c-95c1-661f449a51c5)
![IMG_7047](https://github.com/open-webui/open-webui/assets/47579015/5eacc1c5-17a4-4c10-b5fe-919f2e3aabe4)
![IMG_7493 (2)](https://github.com/open-webui/open-webui/assets/47579015/a4cf1b58-e1f9-463b-aed9-75fe1772dce5)
![IMG_7500](https://github.com/open-webui/open-webui/assets/47579015/0c3d5872-b179-4109-a207-a15da0c7eee5)

**Comment by justinh-rahb** [comment]
I appreciate the setup you have and the clear enthusiasm from everyone involved! If a self-signed certificate will suffice, there are numerous ways to implement it. Our decision not to explain this step is not intended to withhold information or create barriers, but rather the variety of approaches makes it challenging to provide a succinct yet comprehensive explanation that doesn't invite more questions than it answers.

In fact, this would be an excellent opportunity for your students and yourself to delve into and gain a thorough understanding of the technology. Exploring this topic further will not only enable you to implement the technology effectively in a local deployment but also help develop a deeper appreciation for the importance of security measures when working with self-signed certificates and HTTPS.

**Comment by ndonathan** [comment]
I'm going to start this by saying that if you are asking this kind of question, do NOT expose this service to the internet. With that out of the way, There are two methods I suggest in getting this service behind a TLS certificate: 1) NGINX Proxy Manager - This gives you a nice GUI and even the option to get a proper certificate from Let's Encrypt. 2) If you are also needing remote access, I'd get Tailscale setup.

I have used both and NPM is pretty easy if you also have a good understanding of DNS and running your own domain. Tailscale can assist with their Magic DNS feature and making ts really easy to create an exposed service within the VPN with a trusted certificate.

**Reply by justinh-rahb** [reply]
> I'm going to start this by saying that if you are asking this kind of question, do NOT expose this service to the internet.

Quoting this for emphasis. Sorry to say but if you need to ask, you probably _shouldn't_.

**Reply by tbendien** [reply]
Thanks for the Pro Tips 👍 I'm just looking at adding https for my lab/internal network demo environments right now.

For a web facing deployment I would set up something like Keycloak with MFA over Cloudflare. Will try NPM in my lab. Thanks again all!

**Reply by justinh-rahb** [reply]
> Thanks for the Pro Tips 👍 I'm just looking at adding https for my lab/internal network demo environments right now.
> 
> For a web facing deployment I would set up something like Keycloak with MFA over Cloudflare. Will try NPM in my lab. Thanks again all!

To put it simply, I deliberately avoid writing any guides for using solutions like NPM or an Nginx/Letsencrypt setup here, even though they're what I personally use. The reason is that if you haven't set up DNS validation with an API key for your DNS provider, there's no way to obtain a trusted certificate without opening port 80 of your server to the internet. Moreover, I don't bring up self-signed certificates because that raises the issue of "how do I make the certificate validate on my clients?" Once you start going down this path, it can quickly turn into a never-ending rabbit hole of more questions.

After discussing it with Tim and the other contributors, this is just not a level of support we are prepared to offer generally. We're not ignorant of the fact that HTTPS can be attained without exposure to the internet, but the complexities involved are manifold and we don't want to get into that.

That all being said, we won't stop anyone posting their own guides, but it'll be up to the reader to use their own judgment in following them. And just saying... this *is* an open source project. Nothing stopping anybody from having a fork repo with their own Docker image build pipeline that adds in a reverse proxy with SSL termination.

**Comment by shuasimodo** [comment]
Stunnel works pretty great and is bs free. 
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/security_guide/sec-using_stunnel

**Comment by kjackson23** [comment]
I needed to do the same thing and am not familiar with uvicorn server.    I just needed SSL for open-webui on my local network like for any other web application (local or not).  I found the uvicorn documentation here: [https://www.uvicorn.org/deployment/](https://www.uvicorn.org/deployment/).

In open-webui/backend/start.sh add the following to the end of the line that contains "uvicorn".  Currently it is the last line in start.sh, but that may change.

`--ssl-certfile="/path/to/ssl/cert.pem" --ssl-keyfile="/path/to/ssl/privkey.pem"`

Make sure the user that is running start.sh has access to the .pem files, i.e. don't run as root, and since you're not running as root, the user needs access to the files.  My files are located a directory below the open-webui directory and it works just fine.

Edit: I am running ollama and open-webui on Debian Bookworm stable.  Also, adjust the port in start.sh.  I used 8443 instead of 443 so I didn't have to run open-webui as root or sudo (only root privileged users can run on ports 1024 or lower).

**Comment by hdnh2006** [comment]
Hello community! 

I have created an article on my blog talking about how to deploy Open WebUI in https:
https://henrynavarro.org/deploy-your-own-open-webui-interface-with-https-security-53a6ea2609d7?sk=a5876acd56b44ea60f10f9b13aa24aee

And also I have a YouTube video (in spanish 🇪🇸) talking about it:
https://youtu.be/CcGrI9YkUJI?si=d-cskZZTwEdz8Qsv

I hope this could be useful for you.

Enjoy!
